packages/i18next/src/i18nextPlugin.ts

```diff
@@ -0,0 +1,134 @@
+
+    registerInstance(key: string, instance: i18n) {
+        this.#registry.add(key, instance);
+
+        this.#runtime?.logger.debug(`[squide] Registered a new i18next instance with key "${key}":`, instance);
+    }
+
+    getInstance(key: string) {
+        const instance = this.#registry.getInstance(key);
+
+        if (isNil(instance)) {
+            throw new Error(`[squide] Cannot find a registered i18next instance for key: ${key}. Did you forget to register the i18next instance with the i18nextPlugin?`);
+        }
+
+        return instance;
+    }
+
+    detectUserLanguage() {
+        // Could either be detected from a querystring parameter or the user navigator language preferences.
+        let detectedLanguage = this.#languageDetector.detect();
+
+        if (detectedLanguage) {
+            // The navigator language preferences could be something like ["en-US", "en", "fr-CA", "fr"].
+            if (Array.isArray(detectedLanguage)) {
+                this.#runtime?.logger.debug(`[squide] Detected ${detectedLanguage.map(x => `"${x}"`).join(",")} as user language${detectedLanguage.length >= 1 ? "s" : ""}.`);
+
+                // Ensure the navigator language preferences includes at least one supported language.
+                detectedLanguage = findSupportedPreferredLanguage(detectedLanguage, this.#supportedLanguages);
+            } else {
+                this.#runtime?.logger.debug(`[squide] Detected "${detectedLanguage}" as user language.`);
+
+                // Ensure the navigator language preferences includes at least one supported language.
+                detectedLanguage = findSupportedPreferredLanguage([detectedLanguage], this.#supportedLanguages);
+            }
+        }
+
+        if (isNil(detectedLanguage)) {
+            detectedLanguage = this.#fallbackLanguage;
+        }
+
+        this.#currentLanguage = detectedLanguage as T;
+
+        this.#runtime?.logger.debug(`[squide] The language has been set to "${this.#currentLanguage}".`);
+    }
+
+    get currentLanguage() {
+        if (isNil(this.#currentLanguage)) {
+            throw new Error("[squide] The currentLanguage getter is called but no user language has been detected yet. Did you forget to call the detectUserLanguage function?");
+        }
+
+        return this.#currentLanguage;
+    }
+
+    changeLanguage(language: T) {
+        if (!this.#supportedLanguages.includes(language)) {
+            throw new Error(`[squide] Cannot change language for "${language}" because it's not part of the supported languages array. Supported languages are ${this.#supportedLanguages.map(x => `"${x}"`).join(",")}.`);
+        }
+
+        this.#registry.getInstances().forEach(x => {
+            x.changeLanguage(language);
+        });
+
+        this.#currentLanguage = language;
+
+        this.#runtime?.logger.debug(`[squide] The language has been changed to "${this.#currentLanguage}".`);
+    }
+}
+
+export function getI18nextPlugin(runtime: Runtime) {
```

Comment:
shouldn't this method be generic? to forward the property type parameters to i18NextPlugin?

Reply:
Originally I added no casting at all internally because of `i18nextPlugin` genericity. My thinking was that we usually prefer having the consumer make it's own casting for generic type (this is the pattern we agreed on).

Then I had a few cases where it was still useful to return a casted `i18nextPlugin` (without the generic parameter) so I added a basic cast without the generic type. My thinking was that for example `getMswPlugin` as a pluging instance casted as `MswPlugin` so that's consistent.